atc/api/resourceserver/versionserver/downstream.go

```diff
@@ -0,0 +1,56 @@
+package versionserver
+
+import (
+	"encoding/json"
+	"net/http"
+	"strconv"
+
+	"code.cloudfoundry.org/lager"
+
+	"github.com/concourse/concourse/atc/db"
+)
+
+func (s *Server) GetDownstreamResourceCausality(pipeline db.Pipeline) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		logger := s.logger.Session("causality")
+
+		versionIDString := r.FormValue(":resource_config_version_id")
+		resourceName := r.FormValue(":resource_name")
+		versionID, _ := strconv.Atoi(versionIDString)
+
+		resource, found, err := pipeline.Resource(resourceName)
+		if err != nil {
+			logger.Error("failed-to-get-resource", err, lager.Data{"resource-name": resourceName})
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+		if !found {
+			logger.Info("resource-not-found", lager.Data{"resource-name": resourceName})
+			w.WriteHeader(http.StatusNotFound)
+			return
+		}
+
+		causality, found, err := resource.DownstreamCausality(versionID)
+		if err != nil {
+			logger.Error("failed-to-fetch", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+		if !found {
+			logger.Info("resource-version-not-found", lager.Data{"resource-name": resourceName, "resource-config-version": versionID})
+			w.WriteHeader(http.StatusNotFound)
+			return
+		}
+
+		w.Header().Set("Content-Type", "application/json")
+		err = json.NewEncoder(w).Encode(causality)
+		if err != nil {
+			logger.Error("failed-to-encode", err, lager.Data{"resource-name": resourceName, "resource-config-version": versionID})
```

Comment:
According to the ATC logs, this resource version (https://ci.concourse-ci.org/teams/contributor/pipelines/prs/resources/concourse-pr/causality/939850057/downstream?vars.branch=%22master%22) contains a cycle, causing this encoding step to fail:

```
{"timestamp":"2021-06-08T13:42:14.522065439Z","level":"error","source":"atc","message":"atc.causality.failed-to-encode","data":{"error":"json: unsupported value: encountered a cycle via []*atc.CausalityBuild","resource-config-version":939850057,"resource-name":"concourse-pr","session":"4720233"}}
```

Reply:
Ah, this is caused by the `put` step not generating a new version, it just outputs the same version as the `get` step

Reply:
Ended up flattening the api response so it's no longer a tree. This has the side effect of allowing full cycles in the graph (in addition to the "partial" cycles for resources that bump themselves).